### ISSUE TITLE
Page#selelected_page= : Method to select a page WITHOUT animation

### ISSUE THREAD

**Issue by Fredosixx** [SketchUp/api-issue-tracker#876]
Currently, `Pages#selected_page=page` sets the specified page as if te user would click on its tab, that is with an animation hich may last several seconds.

Could we have a method that bypasses the animation?

This would be easier to manage scenes from the API.

**Comment by DanRathbun** [comment]
> _Could we have a method that bypasses the animation?_

We've been doing this for years already using the current API features.

What we do is save the current state, switch off the scene transitions, and then restore them when done.

```ruby
opts = Sketchup.active_model.options["PageOptions"]
prev = opts["ShowTransition"]
opts["ShowTransition"]= false
#
# Do page tasks here ... loop through pages, etc.
#
opts["ShowTransition"]= prev if opts["ShowTransition"] != prev
```

**Comment by Fredosixx** [comment]
Good workaround, worth knowing.

It does however impact the order of observers `frame_change_observer` and `pages_observer`, which are already very messy (events sent several times and order difficult to understand), which is for instance what you want to both track the changes of existing scenes and the selection of scenes (which apparently triggers a change in content).

**Comment by DanRathbun** [comment]
> _It does however impact the order of observers `frame_change_observer` ..._

Yes, @Eneroth3 described the bug in [Issue 244 : Fire FrameChangeObserver at least twice for all scene changes](https://github.com/SketchUp/api-issue-tracker/issues/244)

In the scenario described however, since it is your code switching off the scene transitions, then the current page's properties can be cached before hand. You'd need to add the functionality to pass that data into a frame change observer.

> ... _and `pages_observer`_, ...

I'm assuming that [`#onActiveSectionPlaneChanged`](https://ruby.sketchup.com/Sketchup/EntitiesObserver.html#onActiveSectionPlaneChanged-instance_method) would be most likely affected.

What other [`PagesObsever`](https://ruby.sketchup.com/Sketchup/PagesObserver.html) callbacks would fire when changing a scene ?

----

Regarding a new non-animate scene page changer, it is big no-no in Ruby to have more than 1 argument for an assignment method that uses "=".

So it would have to be named like: `#set_selected_page( page, animate: false )` (or similar.)

**Comment by Fredosixx** [comment]
For info,

when you change a page, the FrameChange observer fires with the percent of animation.

But in addition, if you have the Pages observer on, you get 2 `onContentsModified` events, whereas you are just changing the current scene. These events fire during the animation. I can identify them because the Frame Change with `percent_done` 0 has been received before. Now with the suppression of animation, I get the `onContentsModified` notifications before I get the Frame Change with `percentdone` = 1. So, difficult to indentify a true change in a page versus a simple change of page. This is the side effect.

And indeed, the Pages observer is totally silent when you change the flags of usage (camera, style, layers, etc...). NO event at all. I think this has already been logged in the issue database.

**Comment by DanRathbun** [comment]
> _But in addition, if you have the Pages observer on, you get 2 `onContentsModified` events, whereas you are just changing the current scene. ... So, difficult to identify a true change in a page versus a simple change of page. This is the side effect._

Yep ... Thomas has detailed these internal problems in this request for a new [`PagesObserver`](https://ruby.sketchup.com/Sketchup/PagesObserver.html) callback **4 years** ago ...

* [Issue #237 : Add observer event to detect active scene change #237](https://github.com/SketchUp/api-issue-tracker/issues/237)

> _And indeed, the Pages observer is totally silent when you change the flags of usage (camera, style, layers, etc...). NO event at all. I think this has already been logged in the issue database._

Yea. I logged this request **5 years** ago ...
* [Issue #26 : PagesObserver#onPageUpdate() observer callback #26](https://github.com/SketchUp/api-issue-tracker/issues/26)

**Comment by Fredosixx** [comment]
Thanks. I'll look back at the status in 6 years then....

**Comment by DanRathbun** [comment]
**FYI**: I listed the quirkiness of [`PagesObserver#onContentsModified`](https://ruby.sketchup.com/Sketchup/PagesObserver.html#onContentsModified-instance_method) in this post ... https://github.com/SketchUp/api-issue-tracker/issues/26#issuecomment-691744240

**Comment by thomthom** [comment]
> Could we have a method that bypasses the animation?

Can you elaborate on why you need the API to change the active scene without using the scene's transition properties?

**Comment by Fredosixx** [comment]
to speed time.

**Comment by thomthom** [comment]
Why is that needed for your extension? What is the background for this? What is the extension doing at a higher level?

We need to understand the context of feature requests that we beyond just parity with the UI.

**Comment by Fredosixx** [comment]
The plugin uses a list of scenes to manage extra properties and the user can scroll it.

Dan's trick is OK for removing the animation, but then you have to fix the the mess with the observers.

**Comment by DanRathbun** [comment]
> @thomthom _Can you elaborate on why you need the API to change the active scene without using the scene's transition properties?_

The need centers around the fact that the [`Style`](https://ruby.sketchup.com/Sketchup/Style.html) class is not fully exposed in it's own right.
Instead we must access a scene page's style settings via the [`RenderingOptions`](https://ruby.sketchup.com/Sketchup/RenderingOptions.html) object.

But we cannot access or change the options independently from the current model options. Contrary to what the API docs for [`Page#rendering_options`](https://ruby.sketchup.com/Sketchup/Page.html#rendering_options-instance_method) implies, getting this object and making changes does not work. (I think that there is an open issue on this.)

Instead, in order to make scene page style changes, we must switch to a page via [`Pages#selected_page=`](https://ruby.sketchup.com/Sketchup/Pages.html#selected_page=-instance_method) then access the [model's rendering options](https://ruby.sketchup.com/Sketchup/Model.html#rendering_options-instance_method) and make changes to this, then be sure that the page is set to remember styles or rendering options (one of them may not work as advertised) .. anyway then we must do an [update on the page](https://ruby.sketchup.com/Sketchup/Page.html#update-instance_method) with the correct update flags.

I think that this also is similar to changing the camera for a page. The page to be changed must be the current selected page, the camera changed, then the page updated.

If an extension is doing some of these updates to many or all pages, then a loop must be wrapped within a block like I've shown [above](https://github.com/SketchUp/api-issue-tracker/issues/876#issuecomment-1466810693). Otherwise some properties such as some style and rendering options are interpolated during the animation.

If an extension was going to write out images of all scene page views to disk, the result might be somewhere in the midst of the transition with a mix of styling between the start page and target page.

Anyway, the example of temporarily switching off scene page transitions is not new. We've shown it in the forums for years as SketchUp coding newbs encounter the challenges described in this issue thread.

**Comment by DanRathbun** [comment]
Not sure where or if I remember correctly, but perhaps we've discussed a "**quick iterator**" for the [`Pages`](https://ruby.sketchup.com/Sketchup/Pages.html) collection class like:
```ruby
pages = Sketchup.active_model.pages
pages.cycle do |page|
  # page would now be the selected page after instant transition
end
```
----

An extension or author library could use a wrapper ...
```ruby
    # Cycle through model pages with scene transitions off.
    # Code is responsible to wrap in undo operation if needed.
    #
    # @param model [Sketchup::Model] Defaults to the active model.
    # @yield page [Sketchup::Page] Each page in turn.
    # @return [nil]
    def cycle_pages(model = Sketchup.active_model)
      unless block_given?
        fail(ArgumentError,'Block required',caller)
      end
      opts = model.options["PageOptions"]
      # Save the current page:
      pages = model.pages
      old_page = pages.selected_page
      # Save the current global scene transition setting:
      prev = opts["ShowTransition"]
      # Switch off the global scene transition setting:
      opts["ShowTransition"]= false
        #
        pages.each do |page|
          pages.selected_page= page
          # page should now be the selected page after instant transition
          yield page # Do page tasks in block
        end
        #
        # Restore the original selected page:
        pages.selected_page= old_page
        # Restore the global scene transition setting:
      opts["ShowTransition"]= prev if opts["ShowTransition"] != prev
      # No particular return value:
      nil
    end
```
----
A [refinement](https://ruby-doc.org/core-2.7.2/doc/syntax/refinements_rdoc.html) could be used ...
```ruby
module SomeAuthor
  module PagesCycleRefinement

    refine ::Sketchup::Pages do

      # Cycle through model pages with scene transitions off.
      # Code is responsible to wrap in undo operation if needed.
      #
      # @yield page [Sketchup::Page] Each page in turn.
      # @return [nil]
      def cycle
        unless block_given?
          fail(ArgumentError,'Block required',caller)
        end
        opts = self.model.options["PageOptions"]
        # Save the current page:
        old_page = self.selected_page
        # Save the current global scene transition setting:
        prev = opts["ShowTransition"]
        # Switch off the global scene transition setting:
        opts["ShowTransition"]= false
          #
          self.each do |page|
            self.selected_page= page
            # page should now be the selected page after instant transition
            yield page # Do page tasks in block
          end
          #
          # Restore the original selected page:
          self.selected_page= old_page
          # Restore the global scene transition setting:
        opts["ShowTransition"]= prev if opts["ShowTransition"] != prev
        # Return the collection object as this is not a filter:
        self
      end

    end # class refinement

  end # refinement module
end # top level namespace
```

**Comment by thomthom** [comment]
> The need centers around the fact that the [Style](https://ruby.sketchup.com/Sketchup/Style.html) class is not fully exposed in it's own right.
Instead we must access a scene page's style settings via the [RenderingOptions](https://ruby.sketchup.com/Sketchup/RenderingOptions.html) object.

Are there any use cases outside of that? Sounds to me the real fix is to address the root cause, allowing these properties to be set without the workaround of temporarily changing active page.

I'd favor #878 over this one. We don't want to add API to enable workaround workflow, but rather focus on root cause fixes. (This is why background and context to a feature request is important.)

**Comment by DanRathbun** [comment]
> Are there any use cases outside of that?

**YES.**

Other scenarios in which we need to [quickly cycle through the pages collection](https://github.com/SketchUp/api-issue-tracker/issues/876#issuecomment-1471168521) is to:
* Write out external output files, either thumbnails or full resolution images to a set of external files.
    * Note that the export animation has either not been exposed or was removed. 
  See Issue #875.
    * Regardless of `#875`,  coders still often need to write out scene thumbnails or large images in ways that the export animation command might not cover.

* Gather information from the model relative to each of the scene views, ie:
    * Perhaps shading information for a shadow study
    * Perhaps what is visible or hidden from each particular camera position (_aka_ sightline study.)
 
I could likely search through the forums on `["ShowTransition"]= false` and find more scenarios over the years where we've had to instruct coders to switch off the page transitions to prevent unwanted strangeness.

The main quirk to avoid was that often the output would be somewhere in the midst of a transition, where the rendered view was not at the desired camera position and the render was some interpolated mix between the style of the source page and the target page. This could also result in say a section cut in the wrong location if [`use_section_planes=`](https://ruby.sketchup.com/Sketchup/Page.html#use_section_planes=-instance_method) was set `true`.

Now, again when we speak of this, we always use the term "**quick**" because most often we want to do some batch processing as quickly as possible. So this is why I added the question ([above](https://github.com/SketchUp/api-issue-tracker/issues/876#issuecomment-1471168521)) with examples of a quick `Pages#cycle` iterator.
My question now would be are they the same feature, or should I move the block form `cycle` iterator to a feature request issue of it's own ? I ask because perhaps both ideas may have merit, and the `cycle` iterator would likely leverage the single action page selector.

Keep in mind that an issue that we have even now is that [`Pages#selected_page=`](https://ruby.sketchup.com/Sketchup/Pages.html#selected_page=-instance_method) is (I think) an asynchronous method that mimics the user clicking a scene tab. 
What we coders need is a **non-synchronous** page selector method that does **not** return until the view has **completely** rendered the target page. I am not convinced that the current setter method does this. I seem to recall having to use clunky workarounds involving a temporary [`FrameChangeObserver`](https://ruby.sketchup.com/Sketchup/FrameChangeObserver.html) pattern.

> Sounds to me the real fix is to address the root cause, allowing these properties to be set without the workaround of temporarily changing active page.

I am **not** against a better and direct pattern to change scene page properties 
... and their [axes](https://ruby.sketchup.com/Sketchup/Page.html#axes-instance_method), [camera](https://ruby.sketchup.com/Sketchup/Page.html#camera-instance_method), [rendering_options](https://ruby.sketchup.com/Sketchup/Page.html#rendering_options-instance_method), [shadow_info](https://ruby.sketchup.com/Sketchup/Page.html#shadow_info-instance_method), and [style](https://ruby.sketchup.com/Sketchup/Page.html#style-instance_method).

But these enhancements are more directly covered by Issue #878.

> _I'd favor #878 over this one._

Except that I brought up the API issue with access to scene styles and rendering options **years ago**, (I can't remember if it was public or private,) but Chris responded something like "_(shrug) This is way it is because of the way that the core written_".

This leads me to believe that implementing #878 will be much more complex, costly and time consuming **core changes** than what **this issue** asks for ... a fast non-transition page selector method (and hopefully from that a fast non-transition block form iterator.)

> _We don't want to add API to enable workaround workflow, but rather focus on root cause fixes. (This is why background and context to a feature request is important.)_

Sure, but we are programmers with 15 years experience using the SketchUp API and myself twice that programming seriously with 45 years engineering. We likely know what is broken in the API and what features it does not have that we cannot workaround easily or are so clunky that it diminishes the worth of the API.

I ask that you give us more credit and respect when we ask for API enhancements, as we have most probably already investigated the alternatives.

In my opinion, I should **not** have to waste **hours** defending good requests with these long explanations typed with my arthritic hands.

Fredo's idea for an instantaneous scene change method is **valid** and a **good request**, whose absence has been felt by many SketchUp coders in the past.

I feel that Extensibility Team member often assume wrongly that there must be some easily fixed "root cause bug" that will make API enhancement requests moot. This is **not** going to be the case when requests are made by programmers experienced especially with the SketchUp API. We know what is broken. We know what needs improvement. **Give us a bit more respect, please !**

**P.S.** - If you'd like more explanations & use cases then it may be time to update the issue templates for this tracker, a separate one for bugs, another for enhancements. The latter can make a better query for use case explanation, and you might ask if the request is for a bug workaround.
  * And at the same time fix the misspelling of **reproducible**. Ie, there is no "s" in this word.
  * Fix the misspelling of **specific**. Ie, the only "s" in this word is the first letter.

**Comment by DanRathbun** [comment]
Examples of coders in the past having issues with transitions, needing them disabled whilst change to a page and then doing tasks with that page rendered. Often attempts to workaround [`Sketchup::Pages#selected_page=`](https://ruby.sketchup.com/Sketchup/Pages.html#selected_page=-instance_method) being **asynchronous** using a [`FrameChangeObserver`](https://ruby.sketchup.com/Sketchup/FrameChangeObserver.html) fail as the pattern is clunky, and this observer class has bugs and it's implementation is **weird**.

* **How to wait until scene change is completed**
https://forums.sketchup.com/t/how-to-wait-until-scene-change-is-completed/192983

* **Save monochrome light direction to scene**
Coder need to iterate through several scene pages, setting save for shadow info and setting scene specific north angle.
https://forums.sketchup.com/t/save-monochrome-light-direction-to-scene/198071

* **Update many pages with ruby**
This specific scenario was about setting layer visibilities for scene pages. The coder was iterating the [`Pages`](https://ruby.sketchup.com/Sketchup/Pages.html) collection but erroneously setting the selected page and then the `page.update` pattern. The page transitions were causing issues. Aerilius showed that he should have been using [`Page#set_visibility()`](https://ruby.sketchup.com/Sketchup/Page.html#set_visibility-instance_method) which has been available since v6 and does not require actually transitioning to each page.
https://forums.sketchup.com/t/update-many-pages-with-ruby/49134

**Comment by DanRathbun** [comment]
NOTE: misspelled method name in issue title.
